rhp/v4/server.go

```diff
@@ -0,0 +1,1078 @@
+			Height: cru.State.Index.Height + 1,
+			ID:     cru.Block.ParentID,
+		}
+		if revertedIndex == target {
+			return nil
+		}
+
+		modified := make(map[types.Hash256]types.StateElement)
+		cru.ForEachSiacoinElement(func(sce types.SiacoinElement, created bool, spent bool) {
+			if created {
+				modified[sce.ID] = types.StateElement{
+					ID:        sce.ID,
+					LeafIndex: types.UnassignedLeafIndex,
+				}
+			}
+		})
+
+		for i := range inputs {
+			if se, ok := modified[inputs[i].Parent.ID]; ok {
+				inputs[i].Parent.StateElement = se
+			}
+
+			if inputs[i].Parent.LeafIndex == types.UnassignedLeafIndex {
+				continue
+			} else if inputs[i].Parent.LeafIndex >= cru.State.Elements.NumLeaves {
+				return fmt.Errorf("siacoin input %v is not in the correct state", inputs[i].Parent.ID)
+			}
+			cru.UpdateElementProof(&inputs[i].Parent.StateElement)
+		}
+		base = revertedIndex
+	}
+
+	for _, cau := range applied {
+		if cau.State.Index == target {
+			return nil
+		}
+
+		modified := make(map[types.Hash256]types.StateElement)
+		cau.ForEachSiacoinElement(func(sce types.SiacoinElement, created bool, spent bool) {
+			if created {
+				modified[sce.ID] = sce.StateElement
+			}
+		})
+
+		for i := range inputs {
+			if se, ok := modified[inputs[i].Parent.ID]; ok {
+				inputs[i].Parent.StateElement = se
+			}
+
+			if inputs[i].Parent.LeafIndex == types.UnassignedLeafIndex {
+				continue
+			} else if inputs[i].Parent.LeafIndex >= cau.State.Elements.NumLeaves {
+				return fmt.Errorf("siacoin input %v is not in the correct state", inputs[i].Parent.ID)
+			}
+			cau.UpdateElementProof(&inputs[i].Parent.StateElement)
+		}
+		base = cau.State.Index
+	}
+
+	if base != target {
+		return fmt.Errorf("failed to update basis to target %v, current %v", target, base)
+	}
+	return nil
+}
+
+func (s *Server) handleHostStream(stream net.Conn, log *zap.Logger) {
+	defer stream.Close()
+
+	stream.SetDeadline(time.Now().Add(30 * time.Second)) // set an initial timeout
+	rpcStart := time.Now()
+	id, err := rhp4.ReadID(stream)
+	if err != nil {
+		log.Debug("failed to read RPC ID", zap.Error(err))
+		return
+	}
+	log = log.With(zap.Stringer("rpc", id))
+
+	switch id {
+	case rhp4.RPCSettingsID:
+		err = s.handleRPCSettings(stream)
+	case rhp4.RPCAccountBalanceID:
+		err = s.handleRPCAccountBalance(stream)
+	case rhp4.RPCFormContractID:
+		err = s.handleRPCFormContract(stream)
+	case rhp4.RPCFundAccountsID:
+		err = s.handleRPCFundAccounts(stream)
+	case rhp4.RPCLatestRevisionID:
+		err = s.handleRPCLatestRevision(stream)
+	case rhp4.RPCModifySectorsID:
+		err = s.handleRPCModifySectors(stream)
+	case rhp4.RPCReadSectorID:
+		err = s.handleRPCReadSector(stream)
+	case rhp4.RPCRenewContractID:
+		err = s.handleRPCRenewContract(stream)
+	case rhp4.RPCSectorRootsID:
+		err = s.handleRPCSectorRoots(stream)
+	case rhp4.RPCWriteSectorID:
+		err = s.handleRPCWriteSector(stream)
+	case rhp4.RPCVerifySectorID:
+		err = s.handleRPCVerifySector(stream)
```

Comment:
`map[types.Specifier]func(net.Conn) error`? :)

Reply:
I really dislike `map[types.Specifier]func(net.Conn) error`. It is probably fine in practice and is probably optimized away by Go's compiler, but it looks meh as a replacement for a switch statement when the number of cases is relatively small and known ahead of time. For huge switch statements it can be better for maintenance, but 10 or so cases isn't that.

Reply:
To be clear: I don't mind changing it, but I will grumble about it 😛

Reply:
yeah, fine as is -- I agree it's largely an aesthetic choice. Personally I like having a list of `name -> handler` mappings, one per line, but it *will* be less efficient than a `switch`.